### PR TITLE
Run Node Disk and Volume Validation Checks when Upgrading from Rook to OpenEBS

### PR DESCRIPTION
**Pull request by rrpolanco**
#### What this PR does / why we need it:

Apply node disk validation introduced in https://github.com/replicatedhq/kURL/pull/3674 and volume access modes validation introduced in https://github.com/replicatedhq/pvmigrate/pull/139 when upgrading from Rook to OpenEBS.

#### Which issue(s) this PR fixes:

Fixes #
Internal tracker ref: [sc-61144]

#### Special notes for your reviewer:

Unfortunately, the volume access modes validation cannot run at pre-init time because it requires the `openebs` storage class to be installed.

#### Does this PR introduce a user-facing change?

```release-note
Disk and Volume validation checks will be run prior to migrating from Rook to OpenEBS. A failed validation check will abort the upgrade.
```

### CODE REVIEW

File: addons/openebs/3.3.0/install.sh
Comment (camilamacedo86) [inline addons/openebs/3.3.0/install.sh:81]: Could we create a lib/add to an utils/common file functions such as the following example? So that, every time that we want print a waring that will be yellow, an error will be red and an info blue for example. Also, that can avoid we per mistake forget to reset the color. 

See: 

**Lib example**

```shell
# Turn colors in this script off by setting the NO_COLOR variable in your
# environment to any value:
NO_COLOR=${NO_COLOR:-""}
if [ -z "$NO_COLOR" ]; then
  warning_color=$'\e[1;33m' # required check what would be yellow here
  error_color=$'\e[0;31m'
  reset_color=$'\e[0m'
else
  warning_color=''
  error_color=''
  reset_color=''
fi

function log() { printf '%s\n' "$*"; }
function error() { error_text "ERROR:" $* >&2; }
function fatal() { error "$@"; exit 1; }

function warning_text {
  echo "$warning_color$*$reset_color"
}

function error_text {
  echo "$error_color$*$reset_color"
}

```

**Usage example**

```
 echo -e "my test for the error :\n${value}"

OR 

warning_text "My warning text will be outputted in yellow"
```

File: addons/openebs/3.3.0/install.sh
Comment (camilamacedo86) [inline addons/openebs/3.3.0/install.sh:60]: WDYT about something like rook_to_openebs_migration_checks only?

File: addons/openebs/template/base/install.sh
Comment (camilamacedo86) [inline addons/openebs/template/base/install.sh:60]: This code is duplicated. Could we not centralize it in a common.sh or util.sh under the dir?

File: kurl_util/cmd/pvmigrate/main.go
Comment (camilamacedo86) [inline kurl_util/cmd/pvmigrate/main.go:52]: It seems that we need another subcommand instead of it.

File: kurl_util/cmd/pvmigrate/main.go
Comment (camilamacedo86) [inline kurl_util/cmd/pvmigrate/main.go:78]: same (return the error ideally or use logger.Fatalf() if required`: https://github.com/replicatedhq/kURL/pull/3767/files#r1030680619

File: kurl_util/cmd/pvmigrate/main.go
Comment (camilamacedo86) [inline kurl_util/cmd/pvmigrate/main.go:84]: same (return the error ideally or use logger.Fatalf() if required`: https://github.com/replicatedhq/kURL/pull/3767/files#r1030680619

Also, we can use q+ to print what is returned into `migrate.PrintPVAccessModeErrors(unsupportedPVCs)` formatted.

File: kurl_util/cmd/pvmigrate/main.go
Comment (camilamacedo86) [inline kurl_util/cmd/pvmigrate/main.go:91]: same (return the error ideally or use logger.Fatalf() if required`: https://github.com/replicatedhq/kURL/pull/3767/files#r1030680619

Comment (camilamacedo86) [reply kurl_util/cmd/pvmigrate/main.go:91]: Also, why the \n?

File: addons/openebs/3.3.0/install.sh
Comment (camilamacedo86) [reply addons/openebs/3.3.0/install.sh:81]: @rrpolanco after look it better see that the kURL has an standard/utils to print the warnings already (i.e. https://github.com/replicatedhq/kURL/blob/d8de81118f3a113036d3125733ce806d3421e8c1/addons/rook/1.5.11/install.sh#L43 )

Look:

Screenshot 2022-11-28 at 18 55 42

WDYT about we try to keep here all using the same approaches? So that if we want to improve or change that can be done horizontally.

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: addons/openebs/template/base/install.sh
Comment (emosbaugh) [reply addons/openebs/template/base/install.sh:60]: the add-on version is generated from the template directory. it is duplicated in each add-on version. it is our pattern not to use shared code as we try not to change existing add-on versions to avoid breakage. this is likely a larger conversation if we want to change things at this level.

File: addons/openebs/3.3.0/install.sh
Comment (rrpolanco) [reply addons/openebs/3.3.0/install.sh:65]: There are no labels on the storage class so I can't use that approach.

File: addons/openebs/3.3.0/install.sh
Comment (rrpolanco) [reply addons/openebs/3.3.0/install.sh:81]: Good suggestion. I'll use `logFail()` to output the validation failures from the command.

File: addons/openebs/3.3.0/install.sh
Comment (rrpolanco) [reply addons/openebs/3.3.0/install.sh:60]: The `maybe_` prefix for naming functions is quite prevalent for functions which are contingent up performing some action in this project. I do agree this func needs to a better name, however.

File: kurl_util/cmd/pvmigrate/main.go
Comment (rrpolanco) [reply kurl_util/cmd/pvmigrate/main.go:70]: good suggestion. We want to indicate that when any of the validation checks fail then that is considered a failure by this command and thus we return `os.Exit(10`.

File: kurl_util/cmd/pvmigrate/main.go
Comment (rrpolanco) [reply kurl_util/cmd/pvmigrate/main.go:78]: I've update this.

File: kurl_util/cmd/pvmigrate/main.go
Comment (rrpolanco) [reply kurl_util/cmd/pvmigrate/main.go:91]: Thanks. I changed this to use `logger.Fatalf()`

File: addons/openebs/3.3.0/install.sh
Comment (emosbaugh) [inline addons/openebs/3.3.0/install.sh:60]: https://github.com/replicatedhq/kURL/blob/main/addons/README.md#requirements

> All shell functions defined in the add-on's install.sh script should be prefixed with the name of the add-on to avoid naming collisions, for example function rook_dashboard_ready_spinner().

```suggestion
function openebs_maybe_rook_to_openebs_migration_checks() {
```

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: kurl_util/cmd/pvmigrate/main.go
Comment (laverya) [reply kurl_util/cmd/pvmigrate/main.go:57]: is there a reason NOT to print the version?

We want to be able to figure out the version from the logs a customer sends us showing an error

File: addons/openebs/template/base/install.sh
Comment (camilamacedo86) [inline addons/openebs/template/base/install.sh:67]: Not a blocker 
But could we not use as we have in kURL logStep logWarn logFail as well?

**Review by camilamacedo86 — COMMENTED** [review]
Looking the code all shows fine for me.  🥇  

Just a nit https://github.com/replicatedhq/kURL/pull/3767/files#r1049440977 (not a blocker at all)

I think that would be nice get the lgtm and approval from @ricardomaraschini  before get merged because is who done the implementations of the commands used here

File: addons/openebs/template/base/install.sh
Comment (rrpolanco) [reply addons/openebs/template/base/install.sh:67]: We could. However, mapping messages to different colors is a lot more involved than I expected. I rather we have a separate PR for making the output more CLI UX friendly.

**Comment by rrpolanco** [comment]
> Testgrid Run(s) Executing @ https://testgrid.kurl.sh/run/pr-3767-4df5b37-openebs-3.3.0-k8s-docker-localpv-2022-12-15T14:56:09Z

`localpv migrate from rook` openebs add-on test if failing because it doesn't take into account the installer changes in this PR. I have tested the rook to openebs migration locally and it works.

File: kurl_util/cmd/pvmigrate/main.go
Comment (rrpolanco) [reply kurl_util/cmd/pvmigrate/main.go:57]: @laverya I added this to make the console messages more readable when the installer is doing an upgrade from rook to openebs. Basically, I wanted to seamlessly embed `pvmigrate` in the installer and abstract its usage.

Unfortunately, kURL console logs are rarely provided by the customer.

Comment (rrpolanco) [reply kurl_util/cmd/pvmigrate/main.go:57]: @laverya I will push a change to print the `pvmigrate` change.

**Review by laverya — DISMISSED** [review]

File: addons/openebs/3.3.0/install.sh
Comment (rrpolanco) [reply addons/openebs/3.3.0/install.sh]: I've addressed this.

File: addons/openebs/template/base/install.sh
Comment (emosbaugh) [inline addons/openebs/template/base/install.sh:76]: this may overwrite a failed validation with a successful one since you are running it in a loop

File: addons/openebs/template/base/install.sh
Comment (emosbaugh) [inline addons/openebs/template/base/install.sh:64]:
```suggestion
    local rook_scs
    local rook_default_sc
    rook_scs=$(kubectl get storageclass | grep rook | grep -v '(default)' | awk '{ print $1}') # any non-default rook StorageClasses
    rook_default_sc=$(kubectl get storageclass | grep rook | grep '(default)' | awk '{ print $1}') # any default rook StorageClasses
```

File: kurl_util/cmd/pvmigrate/main.go
Comment (emosbaugh) [inline kurl_util/cmd/pvmigrate/main.go:68]:
```suggestion
		logger.Fatalf("failed to get config: %s", err)
```

File: kurl_util/cmd/pvmigrate/main.go
Comment (emosbaugh) [inline kurl_util/cmd/pvmigrate/main.go:73]:
```suggestion
		logger.Fatalf("failed to create kubernetes clientset: %s", err)
```

File: kurl_util/cmd/pvmigrate/main.go
Comment (emosbaugh) [inline kurl_util/cmd/pvmigrate/main.go:97]:
```suggestion
			logger.Fatal(err)
```

File: kurl_util/cmd/pvmigrate/main.go
Comment (emosbaugh) [inline kurl_util/cmd/pvmigrate/main.go:53]: This will get output as part of [these](https://github.com/replicatedhq/kURL/pull/3767/files#diff-a639b8e65bf2b4327bfb8a263798cad88f2c558c76e1dda96169edb3b7bcf4eeR85-R86) log lines. This should go to stderr using the logger.

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: addons/openebs/template/base/install.sh
Comment (rrpolanco) [reply addons/openebs/template/base/install.sh:76]: Good point. I'll break out of the loop if the command fails and continue.

File: addons/openebs/template/base/install.sh
Comment (rrpolanco) [reply addons/openebs/template/base/install.sh:64]: curious, why is this better? I'd rather have two lines of code vs four.

File: kurl_util/cmd/pvmigrate/main.go
Comment (rrpolanco) [reply kurl_util/cmd/pvmigrate/main.go:53]: why do you think the version info belongs in stdout?

File: kurl_util/cmd/pvmigrate/main.go
Comment (rrpolanco) [reply kurl_util/cmd/pvmigrate/main.go:68]: Fixed.

File: kurl_util/cmd/pvmigrate/main.go
Comment (rrpolanco) [reply kurl_util/cmd/pvmigrate/main.go:73]: Fixed.

File: kurl_util/cmd/pvmigrate/main.go
Comment (rrpolanco) [reply kurl_util/cmd/pvmigrate/main.go:97]: Fixed.

File: kurl_util/cmd/pvmigrate/main.go
Comment (emosbaugh) [reply kurl_util/cmd/pvmigrate/main.go:53]: i think it belongs in stderr but you have it going to stdout. in fact i would say it belongs in stdout if printVersion == true otherwise stderr

File: addons/openebs/template/base/install.sh
Comment (emosbaugh) [reply addons/openebs/template/base/install.sh:64]: trying not to pollute the global namespace

Comment (rrpolanco) [reply addons/openebs/template/base/install.sh:64]: ah, good point.

File: kurl_util/cmd/pvmigrate/main.go
Comment (emosbaugh) [inline kurl_util/cmd/pvmigrate/main.go:54]: this should go to the same as the line above

File: addons/openebs/template/base/install.sh
Comment (emosbaugh) [inline addons/openebs/template/base/install.sh:79]: This will not work. let me see if i can find an example in the code where we have similar logic

```
$ bash test.sh
test.sh: line 5: adsfadsf: command not found
0
$ cat test.sh
#!/bin/bash

set -e

rook_scs_pvmigrate_dryrun_output=$(adsfadsf || true)
echo $?
```

File: addons/openebs/template/base/install.sh
Comment (emosbaugh) [inline addons/openebs/template/base/install.sh:81]: i think this will work but please test it

```suggestion
        if ! rook_scs_pvmigrate_dryrun_output=$($BIN_PVMIGRATE --source-sc "$rook_sc" --dest-sc "$OPENEBS_LOCALPV_STORAGE_CLASS" --rsync-image "$KURL_UTIL_IMAGE" --preflight-validation-only) ; then
            break
        fi
```

**Review by emosbaugh — CHANGES_REQUESTED** [review]

File: addons/openebs/template/base/install.sh
Comment (rrpolanco) [reply addons/openebs/template/base/install.sh:79]: I see. It's the `|| true` that's coercing a return code of `0`.

File: addons/openebs/template/base/install.sh
Comment (rrpolanco) [reply addons/openebs/template/base/install.sh]: This works!

**Review by emosbaugh — DISMISSED** [review]

**Review by emosbaugh — APPROVED** [review]

**Pull request merged**
